File: apps/runtime-worker/src/schemas/execution.py
"""Canonical runtime execution schemas.

These models are the single Python runtime contract used by:

- POST /api/v1/execute
- ExecutionService
- f1_runtime result adaptation

Cross-language TS/Python drift is intentionally handled by issue #145.
Until that is closed, this file must not be treated as proof that contract
sync is solved.
"""
from __future__ import annotations

from enum import StrEnum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel


class OctoModel(BaseModel):
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        extra="forbid",
        str_strip_whitespace=True,
    )


class ExecutionStatus(StrEnum):
    PENDING = "pending"
    QUEUED = "queued"
    DISPATCHED = "dispatched"
    RUNNING = "running"
    WAITING_TOOL = "waiting_tool"
    WAITING_HUMAN = "waiting_human"
    RETRYING = "retrying"
    RETRY_SCHEDULED = "retry_scheduled"
    SUSPENDED = "suspended"
    RECLAIMABLE = "reclaimable"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class LLMConfigSchema(OctoModel):
    primary: str = "litellm/default"
    fallback: list[str] = Field(default_factory=list)
    temperature: float = Field(default=0.7, ge=0.0, le=2.0)
    max_tokens: int | None = Field(default=None, ge=1)
    policy_ref: str | None = None
    provider: str | None = None


class ExecutionLimitsSchema(OctoModel):
    max_usd_per_run: float | None = Field(default=None, ge=0)
    max_tokens_per_run: int | None = Field(default=None, ge=1)
    max_tool_rounds_per_run: int | None = Field(default=None, ge=1)
    max_delegation_depth: int | None = Field(default=None, ge=0)
    max_concurrent_runs: int | None = Field(default=None, ge=1)
    run_timeout_secs: int | None = Field(default=300, ge=30)


class ExecutionRequest(OctoModel):
    execution_id: str
    tenant_id: str
    agent_id: str
    workspace_id: str = "default"
    task: str = Field(default="execute persisted task", min_length=1, max_length=32_000)
    context: dict[str, Any] = Field(default_factory=dict)
    llm: LLMConfigSchema = Field(default_factory=LLMConfigSchema)
    limits: ExecutionLimitsSchema = Field(default_factory=ExecutionLimitsSchema)
    tools: list[str] = Field(default_factory=list)
    streaming: bool = False
    trace_id: str
    run_id: str = ""


class ExecutionResult(OctoModel):
    execution_id: str
    status: ExecutionStatus
    output: str | None = None
    tool_calls: list[dict[str, Any]] = Field(default_factory=list)
    usage: dict[str, int] = Field(default_factory=dict)
    error: str | None = None
    duration_ms: int = Field(ge=0)
    checkpoint: dict[str, Any] | None = None

File: packages/contracts/src/execution.ts
export const ExecutionStatus = {
  PENDING: 'pending',
  QUEUED: 'queued',
  DISPATCHED: 'dispatched',
  RUNNING: 'running',
  WAITING_TOOL: 'waiting_tool',
  WAITING_HUMAN: 'waiting_human',
  RETRYING: 'retrying',
  RETRY_SCHEDULED: 'retry_scheduled',
  SUSPENDED: 'suspended',
  RECLAIMABLE: 'reclaimable',
  COMPLETED: 'completed',
  FAILED: 'failed',
  CANCELLED: 'cancelled',
} as const;

export type ExecutionStatus = (typeof ExecutionStatus)[keyof typeof ExecutionStatus];
export const ExecutionStatusValues = Object.values(ExecutionStatus) as ExecutionStatus[];

export const VALID_TRANSITIONS: Readonly<Record<ExecutionStatus, ReadonlySet<ExecutionStatus>>> = {
  pending: new Set(['queued', 'cancelled']),
  queued: new Set(['dispatched', 'cancelled', 'failed']),
  dispatched: new Set(['running', 'cancelled', 'failed']),
  running: new Set([
    'waiting_tool',
    'waiting_human',
    'retrying',
    'reclaimable',
    'suspended',
    'completed',
    'failed',
    'cancelled',
  ]),
  waiting_tool: new Set(['running', 'retrying', 'failed', 'cancelled', 'suspended']),
  waiting_human: new Set(['running', 'cancelled', 'suspended']),
  retrying: new Set(['retry_scheduled', 'failed', 'cancelled']),
  retry_scheduled: new Set(['queued', 'failed', 'cancelled']),
  reclaimable: new Set([
    'dispatched', // canonical F1 reclaim replay handoff
    'failed',
    'cancelled',
  ]),
  suspended: new Set(['queued', 'cancelled']),
  completed: new Set<ExecutionStatus>(),
  failed: new Set<ExecutionStatus>(),
  cancelled: new Set<ExecutionStatus>(),
};

export const TERMINAL_STATUSES: ReadonlySet<ExecutionStatus> = new Set([
  'completed',
  'failed',
  'cancelled',
]);

export function isTerminalStatus(status: ExecutionStatus): boolean {
  return VALID_TRANSITIONS[status].size === 0;
}

export function isActiveStatus(status: ExecutionStatus): boolean {
  return status === 'running' || status === 'retrying';
}

export function isBlockedStatus(status: ExecutionStatus): boolean {
  return status === 'waiting_tool' || status === 'waiting_human' || status === 'suspended';
}

export class ExecutionTransitionError extends Error {
  readonly from: ExecutionStatus;
  readonly to: ExecutionStatus;
  readonly executionId: string;

  constructor(executionId: string, from: ExecutionStatus, to: ExecutionStatus) {
    super(
      `Invalid execution transition [${executionId}]: ${from} -> ${to}. ` +
        `Valid targets from '${from}': [${[...VALID_TRANSITIONS[from]].join(', ') || 'none'}]`
    );
    this.name = 'ExecutionTransitionError';
    this.from = from;
    this.to = to;
    this.executionId = executionId;
  }
}

export const InvalidTransitionError = ExecutionTransitionError;
export type InvalidTransitionError = ExecutionTransitionError;

export function canTransition(from: ExecutionStatus, to: ExecutionStatus): boolean {
  return VALID_TRANSITIONS[from].has(to);
}

export function assertValidTransition(executionId: string, from: ExecutionStatus, to: ExecutionStatus): void {
  if (!canTransition(from, to)) {
    throw new ExecutionTransitionError(executionId, from, to);
  }
}
